dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/cache/ExpressionDimensionItemAnalyticsCache.java

```diff
@@ -0,0 +1,147 @@
+import org.hisp.dhis.cache.Cache;
+import org.hisp.dhis.cache.CacheProvider;
+import org.springframework.stereotype.Component;
+
+@Slf4j
+@Component
+public class ExpressionDimensionItemAnalyticsCache
+{
+    private final AnalyticsCacheSettings analyticsCacheSettings;
+
+    // Inside the cache there is DataQueryParam.getKey()
+    private final Cache<Set<String>> queryCache;
+
+    /**
+     * Default constructor. Note that a default expiration time is set, as the
+     * TTL will always be overwritten during cache put operations.
+     */
+    public ExpressionDimensionItemAnalyticsCache( CacheProvider cacheProvider,
+        AnalyticsCacheSettings analyticsCacheSettings )
+    {
+        checkNotNull( cacheProvider );
+        checkNotNull( analyticsCacheSettings );
+
+        this.analyticsCacheSettings = analyticsCacheSettings;
+        this.queryCache = cacheProvider.createExpressionDimensionItemAnalyticsResponseCache();
+    }
+
+    /**
+     * This method returns all keys (DataQueryParams::getKey)
+     *
+     * @param key
+     * @return
+     */
+    public Optional<Set<String>> get( String key )
+    {
+        return queryCache.get( key );
+    }
+
+    /**
+     * This method will cache the given DataQueryParams associated with the
+     * given key.
+     * <p>
+     * The TTL of the cached object will be set accordingly to the cache
+     * settings available at {@link AnalyticsCacheSettings}.
+     *
+     * @param key
+     * @param params the DataQueryParams.
+     */
+    public void put( String key, DataQueryParams params )
+    {
+        if ( analyticsCacheSettings.isProgressiveCachingEnabled() )
+        {
+            // Uses the progressive TTL
+            put( key, params,
+                analyticsCacheSettings.progressiveExpirationTimeOrDefault( params.getLatestEndDate() ) );
+        }
+        else
+        {
+            // Respects the fixed (predefined) caching TTL
+            put( key, params, analyticsCacheSettings.fixedExpirationTimeOrDefault() );
+        }
+    }
+
+    /**
+     * Will cache the given key/DataQueryParams pair respecting the TTL provided
+     * through the parameter "ttlInSeconds".
+     *
+     * @param key the cache key associate with the Grid.
+     * @param params the DataQueryParams object to be cached.
+     * @param ttlInSeconds the time to live (expiration time) in seconds.
+     */
+    public void put( String key, DataQueryParams params, long ttlInSeconds )
+    {
+        Optional<Set<String>> dataQueryParamKeySet = queryCache.get( key );
+
+        if ( dataQueryParamKeySet.isPresent() )
+        {
+            dataQueryParamKeySet.get().add( params.getKey() );
+
+            queryCache.put( key, dataQueryParamKeySet.get(), ttlInSeconds );
+        }
+        else
+        {
+            Set<String> paramsSet = new HashSet<>();
+
+            paramsSet.add( params.getKey() );
+
+            queryCache.put( key, paramsSet, ttlInSeconds );
+        }
+    }
+
+    /**
```

Comment:
Missing Javadoc for the `key` argument., which I believe should be renamed to `ediUid`.